proposals/NNNN-mixins.md

```diff
@@ -0,0 +1,501 @@
+
+struct Enemy: MovingObject, AnimatedObject { /*...*/ }
+struct Arrow: MovingObject { /*...*/ }
+```
+
+### Example with logging
+
+`LoggingObject` is not a class of objects in the original meaning. It is more like a feature, or behavior of a data type. Using mixins, we can free the road for other parent classes (or other mixins) in the hierarchy:
+
+```swift
+class Logger { /*...*/ }
+
+mixin LoggingObject {
+  var logger: Logger
+  init(logger: Logger) { self.logger = logger }
+}
+
+class GenericComputer {
+  var result: String? = nil
+  init(options: Options) { /*...*/ }
+  // ...
+}
+
+class SomeComputer: GenericComputer, LoggingObject {
+  init(logger: Logger, options: Options) {
+    LoggingObject.super.init(logger: logger)
+    super.init(options: options)
+  }
+  
+  override func compute() {
+    logger.log("Starting", 1)
+    if someCondition {
+      logger.log("Error", 3)
+    }
+    logger.log("Ending", 1)
+  }
+}
+```
+
+## Detailed design
+
+### Mixin members 
+
+Property requirements:
+
+```swift
+var name: Type { get set }
+```
+
+Property definitions:
+
+```swift
+var name: Type
+```
+
+Method requirements:
+
+```swift
+func name(arg: Type) -> Result
+```
+
+Method definitions:
+
+```swift
+func name(arg: Type) -> Result { }
+```
+
+There is no notion of default methods in mixins. Methods of mixins require `override` to redefine them in types and can be `final`. 
+
+Mixins consider themselves as value types, as protocols and structs. Every member and requirement, which needs to mutate `self`, must be declared as `mutating`.
+
+Subscript requirements and definitions.
+
+Initializer requirements:
+
+```swift
+init(name: Type)
+```
+
+Initializer definitions:
+
+```swift
+init(name: Type) { }
+```
+
+Properties can be immediately initialized: `var name: Type = initializer`
+
+These initialized properties are prepended to all initializers. If no initializers are defined, an implicit default initializer is created.
+
+### Initializers
+
+Mixins can have implemented initializers. They act like initializers of classes. `super` initializer calls are prefixed with parent protocol name for disambiguation purposes. `super` initializer must be explicitly called in initializer of subtype. The syntax for that is `SuperMixinName.super.init()`:
+
+```swift
+mixin A {
+  init(newPath: String) { }
+}
+
+mixin B: A {
+  override init(newPath: String) { }
+}
+```
+
+Default `super` initializers also need to be called explicitly. In multiple inheritance, it is a responsibility of subtype to ensure that initialization order matches semantics of mixins it mixes in:
+
+```swift
+mixin A { }
+mixin B { }
+
+mixin C: A, B {
+  init() {
+    A.super.init()
+    B.super.init()
+  }
+}
+```
+
+Note that `super` syntax can also be used for calling supertype members, as in classes.
+
+### Mixin inheritance
+
+If a single mixin is mixed-in using different paths, then only one copy of it will be kept in the final struct/class. Example:
```

Comment:
I see. Looks like I missed the forest for the trees.

@Evertt Your proposed syntax seems like it could get out of hand when dealing with mixins with many properties. I think the proposal's existing syntax is nearly adequate as-is to express what yours does. I also propose extending the proposal's syntax to allow multiple mixins that share the same base to share the same renamed binding.

```swift
mixin Base { var x: Int }
mixin A: Base {}
mixin B: Base {}
mixin C: Base {}
mixin NotBase { var x: Int }
mixin D: NotBase {}
mixin E: NotBase { var y = NotBase.x } // Is it possible to perform renaming even if there are no conflicts?

struct Q: A, B, C, D {
    var xAB = [A.x, B.x]
    var xAB { A.x B.x } // possible variations on the syntax
    var xC = C.x
    var zDE = [D.x, E.y] // D.x and E.y are both derived from NotBase.x
    // var xAD = [A.x, D.x] // error: A and D aren't derived from the same mixin
}
```
It may also be desirable to give these rename declarations the ability to be marked `override` and/or provide replacement implementations for functions/getters/setters.

The teaching assistant example would be resolved like this:
```swift
mixin UniversityPerson {
    var computerAccount: Account
}
mixin Teacher: UniversityPerson { ... }
mixin Student: UniversityPerson { ... }

class TeachingAssistant: Teacher, Student { 
    var studentAccount = Student.computerAccount
    var teacherAccount = Teacher.computerAccount
    ... 
}
```

I don't agree that the renaming step should be a required part of the mixing-in process if there are no ambiguities. However, I do that that (re)naming mixin members should be mandatory when encountering an ambiguous situation like the diamond inheritance pattern. At a minimum, it would look like this (same mixin declarations as above):
```swift
struct R: A, B {
    var x = [A.x, B.x] // required
}
struct S: A {} // no ambiguities here, no need to (re)name; var x = A.x is implicit
```